src/evm/protocol/uniswap_v4/state.rs

```diff
@@ -282,22 +286,117 @@ impl ProtocolSim for UniswapV4State {
             SimulationError::InvalidInput("I256 overflow: amount_in".to_string(), None)
         })?;
 
-        let result = self.swap(zero_for_one, amount_specified, None)?;
+        let mut amount_to_swap = amount_specified;
+        let mut lp_fee_override: Option<u32> = None;
+        let mut before_swap_gas = 0u64;
+        let mut after_swap_gas = 0u64;
+
+        let token_in_address = Address::from_slice(&token_in.address);
+        let token_out_address = Address::from_slice(&token_out.address);
+
+        let state_context = StateContext {
+            currency_0: if zero_for_one { token_in_address } else { token_out_address },
+            currency_1: if zero_for_one { token_out_address } else { token_in_address },
+            fees: self.fees.clone(),
+            tick: self.tick,
+        };
+
+        let swap_params = SwapParams {
+            zero_for_one,
+            amount_specified,
+            sqrt_price_limit: U256::ZERO, // Will be set to appropriate limit in swap
+        };
+
+        // Check if hook is set and has before_swap permissions
+        if let Some(ref hook) = self.hook {
+            if has_permission(hook.address(), HookOptions::BeforeSwap) {
+                let before_swap_params = BeforeSwapParameters {
+                    context: state_context.clone(),
+                    // TODO is this the right sender? Does this matter?
+                    sender: Address::ZERO,
+                    swap_params: swap_params.clone(),
+                    // TODO what is the hook data?
+                    hook_data: Bytes::new(),
```

Comment:
some hooks take extra parameters as input. This version won't support that type of hooks though. it's ok to leave it empty